File: Iteration_2/src/Grade.java
import java.text.DecimalFormat;

/**
 * The Grade class represents a student's grade with information such as the percentage out of 100,
 * the corresponding Gano, and the letter grade. It provides methods to convert between these
 * representations and perform validations on the input grades.
 *
 * @author mehmet_sina
 * @version 1.0
 * @since 2023-12-10
 */
public class Grade {

    private double outOfHundred;
    private double gano;
    private String letterGrade;

    /**
     * Default constructor for Grade.
     * Creates an instance with default values.
     */
    public Grade() {
        // Default constructor
    }

    /**
     * Constructor for Grade with specified percentage out of 100, Gano, and letter grade.
     *
     * @param outOfHundred The percentage out of 100.
     * @param gano         The Gano equivalent of the grade.
     * @param letterGrade  The letter grade.
     */
    public Grade(double outOfHundred, double gano, String letterGrade) {
        this.outOfHundred = outOfHundred;
        this.gano = gano;
        this.letterGrade = letterGrade;
    }

    /**
     * Retrieves the percentage out of 100.
     *
     * @return The percentage out of 100.
     */
    public double getOutOfHundred() {
        return outOfHundred;
    }

    /**
     * Sets the percentage out of 100.
     *
     * @param outOfHundred The new percentage out of 100 to set.
     */
    public void setOutOfHundred(double outOfHundred) {
        this.outOfHundred = outOfHundred;
    }

    /**
     * Retrieves the Gano equivalent of the grade.
     *
     * @return The Gano equivalent.
     */
    public double getGano() {
        return gano;
    }

    /**
     * Sets the Gano equivalent of the grade.
     *
     * @param gano The new Gano equivalent to set.
     */
    public void setGano(double gano) {
        this.gano = gano;
    }

    /**
     * Retrieves the letter grade.
     *
     * @return The letter grade.
     */
    public String getLetterGrade() {
        return letterGrade;
    }

    /**
     * Sets the letter grade.
     *
     * @param letterGrade The new letter grade to set.
     */
    public void setLetterGrade(String letterGrade) {
        this.letterGrade = letterGrade;
    }

    /**
     * Converts the percentage out of 100 to Gano and returns the numeric Gano value.
     *
     * @param hundredNot The percentage out of 100.
     * @return The numeric Gano value.
     * @throws IllegalArgumentException If the grade is not between 0 and 100.
     */
    public double convertHundredToGano(double hundredNot) {
        if (hundredNot < 0 || hundredNot > 100) {
            throw new IllegalArgumentException("Grade must be between 0 and 100");
        }
        double result = hundredNot / 25;

        DecimalFormat decimalFormat = new DecimalFormat("#.##");
        String formattedResult = decimalFormat.format(result);

        formattedResult = formattedResult.replace(",", "."); // Virgülü nokta ile değiştir
        double numericValue = Double.parseDouble(formattedResult);

        this.gano = Double.parseDouble(formattedResult);
        return numericValue;
    }

    /**
     * Converts the percentage out of 100 to the corresponding letter grade.
     *
     * @param hundred The percentage out of 100.
     * @throws IllegalArgumentException If the grade is not between 0 and 100.
     */
    public void convertHundredToLetterGrade(double hundred) {
        if (hundred < 0 || hundred > 100) {
            throw new IllegalArgumentException("Grade must be between 0 and 100");
        }
        hundred = convertHundredToGano(hundred);
        String letterGrade;
        if (4.00 >= hundred && hundred > 3.50) {
            letterGrade = "AA";
        } else if (3.50 >= hundred && hundred > 3.00) {
            letterGrade = "BA";
        } else if (3.00 >= hundred && hundred > 2.50) {
            letterGrade = "BB";
        } else if (2.50 >= hundred && hundred > 2.00) {
            letterGrade = "CB";
        } else if (2.00 >= hundred && hundred > 1.50) {
            letterGrade = "CC";
        } else if (1.50 >= hundred && hundred > 1.00) {
            letterGrade = "DC";
        } else if (1.00 >= hundred && hundred > 0.50) {
            letterGrade = "DD";
        } else if (0.50 >= hundred && hundred > 0.00) {
            letterGrade = "FD";
        } else {
            letterGrade = "FF";
        }
        this.letterGrade = letterGrade;
    }
}

File: Iteration_3/src/classes/grade.py
class Grade:
    def __init__(
        self, course_code=None, number_grade=None, letter_grade=None, is_passed=None
    ):
        self.course_code = course_code
        self.number_grade = number_grade
        self.letter_grade = letter_grade
        self.is_passed = is_passed

    def get_course_code(self):
        return self.course_code

    def is_grade_passing(self):
        return self.is_passed

    def convert_hundred_to_gano(hundred_not):
        if hundred_not < 0 or hundred_not > 100:
            raise ValueError("Grade must be between 0 and 100")
    
        result = hundred_not / 25
    
        formatted_result = "{:.2f}".format(result)
    
        gano = float(formatted_result.replace(",", "."))
    
        return gano
    
    def convert_hundred_to_letter_grade(self, hundred):
        gano = self.convert_hundred_to_gano(hundred)
        if 4.00 >= gano > 3.50:
            self.letter_grade = "AA"
        elif 3.50 >= gano > 3.00:
            self.letter_grade = "BA"
        elif 3.00 >= gano > 2.50:
            self.letter_grade = "BB"
        elif 2.50 >= gano > 2.00:
            self.letter_grade = "CB"
        elif 2.00 >= gano > 1.50:
            self.letter_grade = "CC"
        elif 1.50 >= gano > 1.00:
            self.letter_grade = "DC"
        elif 1.00 >= gano > 0.50:
            self.letter_grade = "DD"
        elif 0.50 >= gano > 0.00:
            self.letter_grade = "FD"
        else:
            self.letter_grade = "FF"

    def get_info(self):
        return (
            f"course_code: {self.course_code}\n"
            f"number_grade: {self.number_grade}\n"
            f"letter_grade: {self.letter_grade}\n"
            f"is_passed: {self.is_passed}"
        )
